state/src/diff.rs

```diff
@@ -0,0 +1,126 @@
+use {
+    core::fmt,
+    move_core_types::effects::ChangeSet,
+    move_table_extension::{TableChange, TableChangeSet},
+    std::fmt::{Debug, DebugStruct, Formatter},
+};
+
+/// Represents the difference between two [`State`]s.
+///
+/// The difference can be [applied]. If the difference is between two states `S` and `S'`, then it
+/// must be true that `S' := apply(S, diff)`.
+///
+/// `Changes` are usually produced by running a transaction. If `Changes` are produced by running
+/// a transaction on state `S`, then the `S' := apply(S, diff)` represents the state after.
```

Comment:
It would probably be clearer to unify the name of changes in the explanation. E.g. in the second sentence it is mentioned as `Changes` first, but then as `diff`.